File: medium/m_5.py
# https://leetcode.com/problems/longest-palindromic-substring/
# 5. Longest Palindromic Substring

class Solution:
    def longestPalindrome(self, s: str) -> str:
        s = [c for c in s]
        retVal = [s[0]]
        stack = []
        for i in range(1, len(s)):
            
            # Checking for types aa, abba, cbba
            j = i
            stack = s[:i][:] #Converting string to list of chars
            rec = []
            while len(stack) > 0 and j < len(s) and stack.pop() == s[j]:
                rec.append(s[j])
                j = j + 1
            if len(rec) > 0:
                if len(rec[::-1]+ rec) > len(retVal):
                    retVal = rec[::-1]+ rec
                
            # Checking for types aba, ccacc
            j = i
            stack = s[:i][:]
            rec = []
            first = stack.pop()
            while len(stack) > 0 and j < len(s) and stack.pop() == s[j]:
                rec.append(s[j])
                j = j + 1
            if len(rec) > 0:
                if len(rec[::-1]+ [first] + rec) > len(retVal): 
                    retVal = rec[::-1]+ [first] + rec
                
        return "".join(retVal)
            

File: medium/m_5..cs
// https://leetcode.com/problems/longest-palindromic-substring/
// 5. Longest Palindromic Substring

public class Solution {
    public string LongestPalindrome(string s) {
        if(s.Length == 0)
            return string.Empty;
        
        string retVal = string.Empty;
        
        for(int i = 0; i < s.Length; i++){
           string longestOdd = CheckOddPalindrome(s, i);
           string longestEven = CheckEvenPalindrome(s, i);
           if(longestOdd.Length > retVal.Length)
               retVal = longestOdd;
           if(longestEven.Length > retVal.Length)
               retVal = longestEven;
        }
        return retVal;
        
    }
    
    public string CheckOddPalindrome(string s, int i){
        int left = i;
        int right = i;
        while(left >= 0 && right < s.Length && s[left] == s[right]){
            left--;
            right++;
        }
        return s.Substring(left + 1, right - left - 1);
    }
    
    public string CheckEvenPalindrome(string s, int i){
        int left = i;
        int right = i + 1;
        while(left >= 0 && right < s.Length && s[left] == s[right]){
            left --;
            right++;
        }
        return s.Substring(left + 1, right - left - 1);
    }
    
}